fpga.chart.py

```diff
@@ -5,152 +5,161 @@
+    return prefix + '-' + postfix
+
 class Service(SimpleService):
     def __init__(self, configuration=None, name=None):
         SimpleService.__init__(self, configuration=configuration, name=name)
+
         self.conn = None
         self.default_data = dict()
         self.order = []
         self.definitions = dict()
-        self.keys = []
-        self.fpga_mapping = dict()
-        self.next_fpga = 0
-        self.fpga_count = 1
         self.dsn = self.configuration.get('dsn')
-        self.pu_ddr_stats_enable = self.configuration.get('pu_ddr_stats_enable')
-        self.metrics = [ 'bytes', 'jobs', 'max' ]
 
-        conn = self._connect(self.dsn)
-        with conn.cursor() as cursor:
-            cursor.execute('CREATE EXTENSION IF NOT EXISTS swarm64da')
-            cursor.execute('SELECT COUNT(*) FROM swarm64da.get_fpga_stats()')
-            self.fpga_count = cursor.fetchone()[0]
+        self.fpga_mapping = dict()
+        fpga_ids = self.get_fpga_ids()
+        for idx, fpga_id in enumerate(fpga_ids):
+            self.fpga_mapping[fpga_id] = 'fpga-' + str(idx)
+
+        self.fpga_count = len(self.fpga_mapping)
+        self.columns_to_query = self._make_metrics()
+        self.stats_sql = 'SELECT %s FROM swarm64da.get_fpga_stats()' % ','.join(self.columns_to_query)
 
-        if self.fpga_count > 1:
-            for metric in self.metrics:
-                self.init_fpga_metrics(metric, 'fpga-total')
+    def _make_metrics(self):
+        all_metrics = ['bytes', 'jobs', 'max']
+        columns_to_query = set()
 
-        for i in range(self.fpga_count):
-            name = 'fpga-' + str(i)
+        # If more than 1 FPGAs available, add "total" charts for base metrics
+        if self.fpga_count > 1:
+            for metric in all_metrics:
+                self._init_fpga_metric(metric, 'fpga-total', columns_to_query)
 
-            if self.pu_ddr_stats_enable:
-                self.metrics.extend([ 'pu_stats', 'ddr_stats' ])
+        # Detail charts for each FPGA
+        if self.configuration.get('pu_ddr_stats_enable'):
+            all_metrics.extend(['pu_stats', 'ddr_stats'])
 
-            for metric in self.metrics:
-                self.init_fpga_metrics(metric, name)
+        for _, fpga_name in self.fpga_mapping.items():
+            for metric in all_metrics:
+                self._init_fpga_metric(metric, fpga_name, columns_to_query)
 
-        for key in self.keys:
-            self.default_data[key] = 0
+        return ['fpga_id'] + list(columns_to_query)
 
-    def init_fpga_metrics(self, component, name):
-        component_name = name + '-' + component
+    def _init_fpga_metric(self, component, name, columns_to_query):
+        component_name = make_name(name, component)
         component_definition = DEFINITIONS[component]
         self.order.append(component_name)
         self.definitions[component_name] = copy.deepcopy(component_definition)
         self.definitions[component_name]['options'][3] = name
```

Comment:
Why index `3`?

Reply:
The name of the dimension needs to be unique and that is located at index `3`. I will introduce a variable for it.